freqtrade/freqai/base_models/FreqaiMultiOutputClassifier.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+from joblib import Parallel
+from sklearn.base import is_classifier
+from sklearn.multioutput import MultiOutputClassifier, _fit_estimator
+from sklearn.utils.fixes import delayed
+from sklearn.utils.multiclass import check_classification_targets
+from sklearn.utils.validation import check_is_fitted, has_fit_parameter
+
+
+class FreqaiMultiOutputClassifier(MultiOutputClassifier):
+
+    def fit(self, X, y, sample_weight=None, fit_params=None):
+        """Fit the model to data, separately for each output variable.
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            The input data.
+        y : {array-like, sparse matrix} of shape (n_samples, n_outputs)
+            Multi-output targets. An indicator matrix turns on multilabel
+            estimation.
+        sample_weight : array-like of shape (n_samples,), default=None
+            Sample weights. If `None`, then samples are equally weighted.
+            Only supported if the underlying classifier supports sample
+            weights.
+        fit_params : A list of dicts for the fit_params
+            Parameters passed to the ``estimator.fit`` method of each step.
+            Each dict may contain same or different values (e.g. different
+            eval_sets or init_models)
+            .. versionadded:: 0.23
+        Returns
+        -------
+        self : object
+            Returns a fitted instance.
+        """
+
+        if not hasattr(self.estimator, "fit"):
+            raise ValueError("The base estimator should implement a fit method")
+
+        y = self._validate_data(X="no_validation", y=y, multi_output=True)
+
+        if is_classifier(self):
+            check_classification_targets(y)
+
+        if y.ndim == 1:
+            raise ValueError(
+                "y must have at least two dimensions for "
+                "multi-output regression but has only one."
+            )
+
+        if sample_weight is not None and not has_fit_parameter(
+            self.estimator, "sample_weight"
+        ):
+            raise ValueError("Underlying estimator does not support sample weights.")
+
+        if not fit_params:
+            fit_params = [None] * y.shape[1]
+
+        self.estimators_ = Parallel(n_jobs=self.n_jobs)(
```

Comment:
I’d let the base class logic handle this for you. 

Another point here is that this doesn’t necessarily increase time to trained models. It’s adding overhead on Al already highly parallelizable algorithm. Many beta testers ran tests using n_jobs or not, they found no legitimate speed-up. It could be due to using small processors (<=24 cores).


I think the only real usecase for n_jobs is with multiple GPUs installed in the system. Then you’d see tremendous benefit. But this is currently a rare configuration in Freqtrade community, so I wouldn’t prioritize this over the principal benefits of the general PR (multi classification).

Reply:
I copied this from `FreqaiMultiOutputRegressor` directly. My understanding was that this was your solution to `MultiOutputRegressor` not allowing fit_params be passed to underlying models. I must be misunderstanding something?

Reply:
Yes, this is the functionality you duplicated in your `CatboostClassifierMultiTarget.py`. Best to just leave the functionality here in the base class, same as the code you copied this from.

Reply:
Sorry I was mistaken here. You are correct.